sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java

```diff
@@ -1563,7 +1561,7 @@ private class OutgoingData {
         }
       }
 
-      private transient Map<PubsubTopic, OutgoingData> output;
+      private transient Map<KV<PubsubTopic, String>, OutgoingData> output;
```

Comment:
The `PublishRequest` message that's sent to the API specifies a topic and a list of messages.
The limit per request is explicit for the topic field, but the limit of 1 ordering key is only specified in the docs, see https://cloud.google.com/pubsub/docs/ordering#overview_of_message_ordering.

Note to self: Should add a link to the docs.

Reply:
Added links in `PubsubBoundedWriter` and `PubsubUnboundedSink`.

Reply:
Hmmm, are you referring to this sentence in the documentation? 
`To receive a set of ordered messages in a region, you must publish all messages with the same ordering key in the same region.` 

Maybe I'm not understanding it right, but I think the language doesn't explicitly say that a publish request is limited to one ordering key. This line seems to imply this is not the case:
`An ordering key is not equivalent to a partition in a partition-based messaging system as ordering keys are expected to have a much higher cardinality than partitions.`

Might be worth double checking with someone in Pubsub. 
______
P.S. this is not to discourage batching it this way -- it may indeed be the best way. I'd just like to be aware if this is **required**, or if it's otherwise preferred for some reason (e.g. performance, parallelism, etc)

Reply:
I re-read the page and that tiny detail is not mentioned explicitly at all in the public documentation, but attempting to publish multiple messages with multiple ordering keys in a single publish request will result in a validation exception being returned by the Pub/Sub API.